src/dataProvider.ts

```diff
@@ -0,0 +1,181 @@
+import * as http from 'http';
+import fetch, { RequestInit, Response, HeadersInit, BodyInit } from 'node-fetch';
+import { RateLimitError, BadRequestError, InternalServerError } from './errors';
+import delay from './delay';
+import CryptoJS from 'crypto-js';
+import Base64 from 'crypto-js/enc-base64';
+
+
+export enum DatacenterHostnames {
+    USEastCoast = 'use-data.adsrvr.org', // US East Coast
+    UKEU = 'euw-data.adsrvr.org', // UK/EU
+    APAC = 'sin-data.adsrvr.org', // APAC
+    USWestCoast = 'usw-data.adsrvr.org', // US West Coast
+    Tokyo = 'tok-data.adsrvr.org', // Tokyo
+    China = 'https://data-cn2.adsrvr.cn' // China
+}
+
+interface DataProviderOptions {
+    /**
+     * The url of the Tradedesk Data API
+     */
+    apiUrl?: string;
+
+    /**
+     * HTTP(s) Agent to use when making requests
+     */
+    agent?: http.Agent | ((parsedUrl: URL) => http.Agent);
+
+    /**
+     * Max number of retires to try a request when getting rate limited
+     */
+    maxRetries?: number
+
+    /**
+     * Max delay for an exponential back off policy for when retrying requests. Time is in seconds.
+     */
+    maxRetryDelay?: number
+
+    /**
+     * Base delay for an exponential back off policy for when retrying requests. Time is in seconds.
+     */
+    retryDelay?: number
+}
+
+class DataProvider {
+
+     /**
+     * Options used to make requests
+     */
+    public options: DataProviderOptions;
+
+    /**
+     * Provider Provisioned Secret Key
+     */
+    private secretKey: string;
+
+    constructor(secretKey: string, options: DataProviderOptions = {}) {
+        this.secretKey = secretKey
+        this.options = {
+            apiUrl: `https://${DatacenterHostnames.USEastCoast}`,
+            maxRetries: 3,
+            maxRetryDelay: 60,
+            retryDelay: 5,
+            ...options,
+        };
+    }
+
+    /**
+     * Set the API url based on an enumerated list
+     */
+    setApiUrl(dataCenter: DatacenterHostnames): DataProvider;
+
+    /**
+     * Set the API Url to an explicit url
+     */
+    setApiUrl(url: string): DataProvider;
+
+    setApiUrl(arg: any): DataProvider {
+        if (Object.values(DatacenterHostnames).includes(arg)) {
+            this.options.apiUrl = `https://${arg}`;
```

Comment:
I don't understand the reason for this check.

* Get the values of the known hostnames
* Then check if that array includes a known hostname (not the enum key for the value)
* If it does, add `https://` to the front of the `arg` and return

Could this be simplified to `if (arg.indexOf('https://') !== 0) {`?

Reply:
I want to explicitly check that the user was giving me one of the approved domain names before I prepended "https" to it and I don't like checking for uri schemas

Reply:
I just refactored to add 'https' to all DatacenterHostnames enum eliminating the need for the check.